loglist/diff_check.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package loglist
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"reflect"
+  "sort"
+)
+
+// warningCollector is an interface for structs that have the capability to
+// collect warnings.  It is used to report warnings from the various check
+// functions defined in this file.
+type warningCollector interface {
+	addWarning(string)
+}
+
+type warningList struct {
+	warnings []string
+}
+
+func (wl *warningList) addWarning(w string) {
+	if w != "" {
+		wl.warnings = append(wl.warnings, w)
+	}
+}
+
+// Check: operators lists are identical.
+func checkOperators(master_ll *LogList, branch_ll *LogList, wc warningCollector) {
+	if !reflect.DeepEqual(master_ll.Operators, branch_ll.Operators) {
+		wc.addWarning("Operators lists are not identical")
+	}
+}
+
+// Check: 2 logs are identical (apart from description and STH).
+func checkLogPair(log_one *Log, log_two *Log, wc warningCollector) {
+	// Description and STH comparison are omitted.
+
+	if !bytes.Equal(log_one.Key, log_two.Key) {
+		wc.addWarning(fmt.Sprintf(
+			"Log %s and log %s have different keys.",
+			log_one.Description, log_two.Description))
+	}
+	if log_one.MaximumMergeDelay != log_two.MaximumMergeDelay {
+		wc.addWarning(fmt.Sprintf(
+			"Maximum merge delay mismatch for logs %s and %s: %d != %d.",
+			log_one.Description, log_two.Description, log_one.MaximumMergeDelay, log_two.MaximumMergeDelay))
+	}
+	sort.Sort(sort.IntSlice(log_one.OperatedBy))
```

Comment:
You should probably sort a copy of these slices -- this modifies the input values, which might surprise a caller.

Reply:
Done.